src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -15,65 +15,309 @@
+        final int digit3 = input[start + negative + has4 + 2];
+
+        return (-negative ^ (has4 * (digit1 * 100) + digit2 * 10 + digit3 - 528) - negative); // 528 == ('0' * 10 + '0')
+    }
+
+    // branchless max (unprecise for large numbers, but good enough)
+    static int max(final int a, final int b) {
+        final int diff = a - b;
+        final int dsgn = diff >> 31;
+        return a - (diff & dsgn);
+    }
+
+    // branchless min (unprecise for large numbers, but good enough)
+    static int min(final int a, final int b) {
+        final int diff = a - b;
+        final int dsgn = diff >> 31;
+        return b + (diff & dsgn);
+    }
+
+    /**
+     * A normal Java HashMap does all these safety things like boundary checks... we don't need that, we need speeeed.
+     *
+     * So I've written an extremely simple linear probing hashmap that should work well enough.
+     */
+    class BitTwiddledMap {
+        private static final int SIZE = 16384; // A bit larger than the number of keys, needs power of two
+        private int[] indices = new int[SIZE]; // Hashtable is just an int[]
+
+        BitTwiddledMap() {
+            // Optimized fill with -1, fastest method:
+            int len = indices.length;
+            if (len > 0) {
+                indices[0] = -1;
+            }
+            // Value of i will be [1, 2, 4, 8, 16, 32, ..., len]
+            for (int i = 1; i < len; i += i) {
+                System.arraycopy(indices, 0, indices, i, i);
+            }
+        }
+
+        private List<Entry> values = new ArrayList<>(512);
+
+        record Entry(int hash, byte[] key, Measurement measurement) {
+            @Override
+            public String toString() {
+                return new String(key) + "=" + measurement;
+            }
+        }
+
+        /**
+         * Who needs methods like add(), merge(), compute() etc, we need one, getOrCreate.
+         * @param key
+         * @return
+         */
+        public Measurement getOrCreate(byte[] key, int length) {
+            int inHash;
+            int index = (SIZE - 1) & (inHash = hashCode(key, length));
+            int valueIndex;
+            Entry retrievedEntry = null;
+            while ((valueIndex = indices[index]) != -1 && (retrievedEntry = values.get(valueIndex)).hash != inHash) {
+                index = (index + 1) % SIZE;
+            }
+            if (valueIndex >= 0) {
+                return retrievedEntry.measurement;
+            }
+            // New entry, insert into table and return.
+            indices[index] = values.size();
+
+            // Only parse this once:
+            byte[] actualKey = new byte[length];
+            System.arraycopy(key, 0, actualKey, 0, length);
+
+            Entry toAdd = new Entry(inHash, actualKey, new Measurement());
+            values.add(toAdd);
+            return toAdd.measurement;
+        }
+
+        private static int hashCode(byte[] a, int length) {
```

Comment:
Hash code here has a data dependency: you either manually unroll this or just relax the hash code by using a car handle and use getLong amortizing the data dependency in batches, handing only the last 7 (or less) bytes separately, using the array